File: AI_admin_Agent/backend/app/core/langgraph_checkpointer.py
"""LangGraph Checkpointer：memory | postgres（Phase 5）。"""
from __future__ import annotations

import os

from langgraph.checkpoint.memory import MemorySaver

_memory_saver: MemorySaver | None = None
_postgres_saver = None
_postgres_ready = False


def _mode() -> str:
    return os.getenv("ADMIN_LANGGRAPH_CHECKPOINTER", "0").strip().lower()


def is_admin_langgraph_checkpointer_enabled() -> bool:
    return _mode() not in ("0", "false", "off", "")


def init_admin_postgres_checkpointer() -> bool:
    global _postgres_saver, _postgres_ready
    if _mode() not in ("postgres", "pg"):
        return False
    url = (
        os.getenv("AGENT_DATABASE_URL")
        or os.getenv("CLAWHIVE_DATABASE_URL")
        or os.getenv("DATABASE_URL")
        or ""
    ).strip()
    if not url:
        return False
    try:
        from langgraph.checkpoint.postgres import PostgresSaver

        saver = PostgresSaver.from_conn_string(url.replace("postgresql+psycopg2:", "postgresql:"))
        saver.setup()
        _postgres_saver = saver
        _postgres_ready = True
        return True
    except Exception:
        _postgres_saver = None
        _postgres_ready = False
        return False


def get_admin_langgraph_checkpointer():
    global _memory_saver
    if not is_admin_langgraph_checkpointer_enabled():
        return None
    if _mode() in ("postgres", "pg"):
        return _postgres_saver
    if _memory_saver is None:
        _memory_saver = MemorySaver()
    return _memory_saver


def resolve_langgraph_thread_id(session_id: str | None, trace_id: str | None = None) -> str | None:
    if not is_admin_langgraph_checkpointer_enabled():
        return None
    sid = str(session_id or "").strip()
    if _mode() in ("postgres", "pg") and sid:
        return f"sess-{sid}"
    run = str(trace_id or "").strip()
    if run:
        return f"run-{run}"
    return f"sess-{sid}" if sid else None


def build_graph_invoke_config(session_id: str | None, trace_id: str | None = None) -> dict | None:
    tid = resolve_langgraph_thread_id(session_id, trace_id)
    if not tid:
        return None
    return {"configurable": {"thread_id": tid}}


def checkpointer_status() -> dict:
    return {
        "mode": _mode(),
        "enabled": is_admin_langgraph_checkpointer_enabled(),
        "postgresReady": _postgres_ready,
    }

File: Manager_Agent/server/graph/core/runtime/langgraphCheckpointer.ts
import { MemorySaver } from '@langchain/langgraph-checkpoint'
import { resolveAgentDatabaseUrl } from '#agent-shared/storageBackend'

export type CheckpointerMode = 'off' | 'memory' | 'postgres'

let memorySaver: MemorySaver | null = null
let postgresSaver: unknown | null = null
let postgresReady = false
let postgresInitPromise: Promise<void> | null = null

export function resolveManagerCheckpointerMode(env: NodeJS.ProcessEnv = process.env): CheckpointerMode {
  const raw = String(env.MANAGER_LANGGRAPH_CHECKPOINTER ?? '').trim().toLowerCase()
  if (raw === 'postgres' || raw === 'pg') return 'postgres'
  if (raw === '1' || raw === 'true' || raw === 'memory') return 'memory'
  if (raw === '0' || raw === 'false' || raw === 'off') return 'off'
  const url = resolveAgentDatabaseUrl(env)
  if (url && String(env.MGR_CHECKPOINTER_AUTO ?? '1').trim() !== '0') return 'postgres'
  return 'off'
}

export function isManagerLangGraphCheckpointerEnabled(env: NodeJS.ProcessEnv = process.env): boolean {
  return resolveManagerCheckpointerMode(env) !== 'off'
}

export async function initManagerPostgresCheckpointer(env: NodeJS.ProcessEnv = process.env): Promise<boolean> {
  if (resolveManagerCheckpointerMode(env) !== 'postgres') return false
  if (postgresReady && postgresSaver) return true
  if (!postgresInitPromise) {
    postgresInitPromise = (async () => {
      const url = resolveAgentDatabaseUrl(env)
      if (!url) return
      try {
        const mod = await import('@langchain/langgraph-checkpoint-postgres')
        const PostgresSaver = (mod as { PostgresSaver: { fromConnString: (u: string) => { setup: () => Promise<void> } } })
          .PostgresSaver
        const saver = PostgresSaver.fromConnString(url.replace(/^postgresql\+psycopg2:/, 'postgresql:'))
        await saver.setup()
        postgresSaver = saver
        postgresReady = true
      } catch {
        postgresSaver = null
        postgresReady = false
      }
    })()
  }
  await postgresInitPromise
  return Boolean(postgresSaver)
}

export function getManagerLangGraphCheckpointer(): MemorySaver | unknown | undefined {
  const mode = resolveManagerCheckpointerMode()
  if (mode === 'off') return undefined
  if (mode === 'postgres') {
    return postgresSaver ?? undefined
  }
  if (!memorySaver) memorySaver = new MemorySaver()
  return memorySaver
}

export function resetManagerLangGraphCheckpointerForTests(): void {
  memorySaver = null
  postgresSaver = null
  postgresReady = false
  postgresInitPromise = null
}

/** 图 invoke：每轮 run 独立 thread；无 runId 时才回退 sess-*（HITL 续跑等） */
export function resolveLangGraphThreadId(input: {
  runId?: string
  sessionId?: string
  threadId?: string
  /** @deprecated 每轮 run 已默认隔离；保留兼容 */
  freshThread?: boolean
}): string | undefined {
  if (!isManagerLangGraphCheckpointerEnabled()) return undefined
  const run = String(input.runId || '').trim()
  if (run) return `run-${run}`
  const mode = resolveManagerCheckpointerMode()
  const sid = String(input.sessionId || input.threadId || '').trim()
  if (mode === 'postgres' && sid) return `sess-${sid}`
  return sid ? `sess-${sid}` : undefined
}

export function getManagerCheckpointerStatus() {
  return {
    mode: resolveManagerCheckpointerMode(),
    postgresReady,
    enabled: isManagerLangGraphCheckpointerEnabled()
  }
}
